src/posts/2022-11-03-constellation/index.md

```diff
@@ -0,0 +1,148 @@
+4. Cilium-agent watches and consumes all CiliumEndpointSlices
+5. Cilium-agent writes a new entry in IPCache
+
+From this flow, it is obvious that the IPCache is only eventually consistent, and we should be able to reproduce the issue by disabling the Cilium operator.
+
+To deterministically reproduce the issue, we can adapt the WireGuard integration test in [`test/k8s/datapath_configuration.go`](https://github.com/edgelesssys/cilium/blob/d465085df155502267589f782801c63f3d08dd40/test/k8s/datapath_configuration.go#L545).  In this test, the execution is explicitly halted until the endpoint has propagated:
+
+```golang
+// Due to IPCache update delays, it can take up to a few seconds
+// before both nodes have added the new pod IPs to their allowedIPs
+// list, which can cause flakes in CI. Therefore wait for the
+// IPs to be present on both nodes before performing the test
+```
+
+Using this test, we scale the Cilium-operator deployment to 0 before deploying the test daemonset. If we now remove the check for the updated WireGuard peers mentioned above, the tests fail since the new endpoints can’t properly propagate to the IPCache and unencrypted traffic is leaked.
+
+In summary, when network traffic is sent to newly created Kubernetes Pods, a short delay in the propagation of routing information can lead to Cilium sending that traffic unencrypted over the wire.
+ 
+### The strict mode solution
+
+Now that we [have](https://github.com/edgelesssys/cilium/blob/d465085df155502267589f782801c63f3d08dd40/test/k8s/datapath_configuration.go#L631) a test case, we can begin to develop a solution. We call it the strict mode.
+
+With Constellation we are in the unique position that we control the whole underlying stack, including the Kubernetes and Cilium bootstrapping process. Therefore, we can automatically deploy and configure Cilium with our strict mode.
+
+We define the following constraints for our strict mode:
+
+- IPv4 only (Pod network commonly IPv4 only)
+- PodCIDR must be known beforehand
+- Both vxlan (used for AWS/Azure) and native routing (used for GCP) must be supported
+
+While the IPv4 limitation can be easily solved later, the PodCIDR must be static and known before deploying Cilium. Support for a dynamic PodCIDR is a complex feature to implement, because of the same eventual consistency problems we are trying to solve in the first place. Since we support Cilium’s tunneling and direct routing capabilities, the strict mode must also work with both. Note that the test we adapted already covers both scenarios.
+
+> A strict mode for AWS and Azure with direct routing poses some challenges. Since Pod IPs are allocated out of the same subnet as Node IPs, we cannot filter by subnet. Moreover, we cannot rely on the IPCache to add something like an allow-node rule to the filter. The Node could have been terminated and its IP re-used as a Pod IP while the IP is still associated with the Node in the IPCache. For that reason, we’re using vxlan.
```

Comment:
> The Node could have been terminated and its IP re-used as a Pod IP while the IP is still associated with the Node in the IPCache.

That sounds like a bug to me. If you noticed this behavior, could you open an issue with more details? Could it be the same bug as https://github.com/cilium/cilium/issues/21844?

Reply:
I didn't specifically encounter this issue. But since the IPCache is only eventually updated, I imagine there are ways to artificially slow down the propagation by e.g. specifically disrupting traffic between cilium-operator/K8sAPI and cilium-agent. Having said that, I didn't try to create this issue since we could side-step it by simply using vxlan. 

This topic also came up during the initial discussion regarding this topic: https://cilium.slack.com/archives/CDKG8NNHK/p1654789894183409?thread_ts=1654788069.701179&cid=CDKG8NNHK. Did I misunderstand something and this actually constitutes a bug?

If you wish, we could clarify that we have not reproduced this and is merely a theoretical argument. Also, if you have any suggestions on how to make this point more concise and simpler to understand, please let us know :smile:

Reply:
> I didn't specifically encounter this issue. But since the IPCache is only eventually updated, I imagine there are ways to artificially slow down the propagation by e.g. specifically disrupting traffic between cilium-operator/K8sAPI and cilium-agent.

Ah, yes, that is correct because the ipcache is eventually consistent. In practice there is some sort of grace period before IP addresses are reallocated so this doesn't become an issue, but it's probably possible to force it by e.g. overloading the kube-apiserver for long enough that the targeted reallocation happens. But then it would also be hard to perform the attack given kube-apiserver is overloaded.

I think it would still be useful to cover these ENI & Azure cases by relying on the ipcache. We can probably add a caveat section in the documentation to mention the above limitation.

Reply:
Thanks for the input, your proposal sounds good to me. While it is somewhat weird to go from the current state which only offers eventual security to another state which (in some cases) only offers eventual security, I agree that also it has only upside to also support the ENI and Azure case. Then the strict mode is something that goes from a more niche option to "just turn this on in most cases for more security". Moreover, I agree with you that the attack is quite difficult. Also, wrongly sending out unencrypted traffic by chance is highly unlikely for normal operation.

My proposal would be to briefly summarize the solution and its benefits and limitations in place of the commented note.  I'll probably start working on this feature in the next few days.